bin/faucet/src/server/challenge.rs

```diff
@@ -0,0 +1,161 @@
+        Self { difficulty, timestamp, signature }
+    }
+
+    /// Decodes the challenge and verifies that the signature part of the challenge is valid
+    /// in the context of the specified secret.
+    pub fn decode(value: &str, secret: [u8; 32]) -> Result<Self, InvalidRequest> {
+        // Add the "0x" prefix to the hex string, otherwise the hex_to_bytes function will fail.
+        let value = "0x".to_string() + value;
+        // Parse the hex-encoded challenge string
+        let bytes = hex_to_bytes::<48>(&value).map_err(|_| InvalidRequest::MissingPowParameters)?;
+
+        if bytes.len() != 48 {
+            // 8 + 8 + 32 bytes
+            return Err(InvalidRequest::MissingPowParameters);
+        }
+
+        // SAFETY: Length of 48 is enforced above.
+        let difficulty = u64::from_le_bytes(bytes[0..8].try_into().unwrap()) as usize;
+        let timestamp = u64::from_le_bytes(bytes[8..16].try_into().unwrap());
+        let signature: [u8; 32] = bytes[16..48].try_into().unwrap();
+
+        // Verify the signature
+        let expected_signature = Self::compute_signature(secret, difficulty, timestamp);
+        if signature == expected_signature {
+            Ok(Self::from_parts(difficulty, timestamp, signature))
+        } else {
+            Err(InvalidRequest::ServerSignaturesDoNotMatch)
+        }
+    }
+
+    /// Encodes the challenge into a hex string.
+    pub fn encode(&self) -> String {
+        let mut bytes = Vec::with_capacity(48);
+        bytes.extend_from_slice(&(self.difficulty as u64).to_le_bytes());
+        bytes.extend_from_slice(&self.timestamp.to_le_bytes());
+        bytes.extend_from_slice(&self.signature);
+        bytes.to_hex()
+    }
+
+    /// Checks whether the provided nonce satisfies the difficulty requirement encoded in the
+    /// challenge.
+    pub fn validate_pow(&self, nonce: u64) -> bool {
+        let mut hasher = Sha3_256::new();
+        hasher.update(self.encode());
+        hasher.update(nonce.to_le_bytes());
+        let hash = hasher.finalize();
+
+        let leading_zeros = hash.iter().take_while(|&b| *b == 0).count();
+
+        leading_zeros >= self.difficulty
+    }
+
+    /// Checks if the challenge timestamp is expired.
+    pub fn is_expired(&self, current_time: u64) -> bool {
+        (current_time - self.timestamp) > CHALLENGE_LIFETIME_SECONDS
+    }
+
+    /// Computes the signature for a challenge.
+    fn compute_signature(secret: [u8; 32], difficulty: usize, timestamp: u64) -> [u8; 32] {
+        let mut hasher = Sha3_256::new();
+        hasher.update(secret);
+        hasher.update(difficulty.to_le_bytes());
+        hasher.update(timestamp.to_le_bytes());
+        hasher.finalize().into()
+    }
+
+    /// Validates the challenge and nonce.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The challenge is expired.
+    /// * The challenge is invalid.
+    pub fn validate(&self, nonce: u64) -> Result<(), InvalidRequest> {
```

Comment:
I probably wouldn't add this function - or if we do add it, I would:

- Pass the time in as a parameter.
- Use a different error type (not `InvalidRequest` but create a new error enum specifically for `PoW`).

Basically, I'm thinking about this through the lens of moving `PoW` components to the `utils` crate eventually.